websocket.go

```diff
@@ -0,0 +1,277 @@
+		messageBytes, err := json.Marshal(msg)
+		if err != nil {
+			log.Printf("Failed to marshal subscribe message: %v", err)
+			return
+		}
+
+		// Write the message to the upstream WebSocket
+		err = p.upstream.Write(context.Background(), websocket.MessageText, messageBytes)
+		if err != nil {
+			log.Printf("Failed to forward subscribe message to upstream: %v", err)
+		}
+	}
+}
+
+func (p *WebSocketProxy) handleUnsubscribe(conn *websocket.Conn, msg map[string]any) {
+	swapIDs, ok := msg["args"].([]any)
+	if !ok {
+		log.Printf("Invalid unsubscribe message: %v", msg)
+		return
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for _, swapID := range swapIDs {
+		id, ok := swapID.(string)
+		if !ok {
+			log.Printf("Invalid swap ID: %v", swapID)
+			continue
+		}
+
+		// Remove client from swap ID's subscribers
+		delete(p.subscribers[id], conn)
+		if len(p.subscribers[id]) == 0 {
+			delete(p.subscribers, id)
+		}
+
+		// Remove swap ID from client's subscriptions
+		delete(p.clients[conn], id)
+	}
+
+	// Forward unsubscribe message to upstream
+	if p.upstream != nil {
+		// Marshal the message into JSON ([]byte)
+		messageBytes, err := json.Marshal(msg)
+		if err != nil {
+			log.Printf("Failed to marshal unsubscribe message: %v", err)
+			return
+		}
+
+		// Write the message to the upstream WebSocket
+		err = p.upstream.Write(context.Background(), websocket.MessageText, messageBytes)
+		if err != nil {
+			log.Printf("Failed to forward unsubscribe message to upstream: %v", err)
+		}
+	}
+}
+
+func (p *WebSocketProxy) Start() {
+	// Connect to upstream WebSocket server
+	err := p.connectToUpstream()
+	if err != nil {
+		log.Fatalf("Failed to connect to upstream WebSocket: %v", err)
+	}
+
+	// Start a goroutine to listen for messages from the upstream server
+	go func() {
+		for {
+			_, message, err := p.upstream.Read(context.Background())
+			if err != nil {
+				log.Printf("Upstream WebSocket read error: %v", err)
+				break
+			}
+
+			// Parse the message
+			var msg map[string]any
+			if err := json.Unmarshal(message, &msg); err != nil {
+				log.Printf("Failed to parse upstream message: %v", err)
+				continue
+			}
+
+			// Extract swap IDs from the message
+			args, ok := msg["args"].([]any)
+			if !ok {
+				log.Printf("Invalid upstream message: %v", msg)
+				continue
+			}
+
+			// Broadcast message to clients subscribed to the swap IDs
+			p.mu.Lock()
+			for _, arg := range args {
+				swap, ok := arg.(map[string]any)
+				if !ok {
+					continue
+				}
+				swapID, ok := swap["id"].(string)
+				if !ok {
+					continue
+				}
+
+				for client := range p.subscribers[swapID] {
+					// Marshal the message into JSON ([]byte)
+					messageBytes, err := json.Marshal(msg)
+					if err != nil {
+						log.Printf("Failed to marshal message for client: %v", err)
+						continue
+					}
+
+					// Write the message to the client WebSocket
+					err = client.Write(context.Background(), websocket.MessageText, messageBytes)
```

Comment:
Ideally writes are done outside of the main lock. You can do that by only taking the lock when getting the subscriber from the map and when deleting the subscriber.